OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs

```diff
@@ -0,0 +1,145 @@
+using OutOfSchool.Services.Models;
+using OutOfSchool.WebApi.Models;
+using OutOfSchool.WebApi.Services;
+
+namespace OutOfSchool.WebApi.Controllers
+{
+    [ApiController]
+    [Route("[controller]/[action]")]
+    [Authorize(AuthenticationSchemes = "Bearer")]
+    public class ProviderController : ControllerBase
+    {
+        private readonly ILogger<ProviderController> logger;
+        private readonly IProviderService providerService;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ProviderController"/> class.
+        /// </summary>
+        /// <param name="logger">Logging mechanism.</param>
+        /// <param name="providerService">Service for Provider model.</param>
+        public ProviderController(ILogger<ProviderController> logger, IProviderService providerService)
+        {
+            this.logger = logger;
+            this.providerService = providerService;
+        }
+
+        /// <summary>
+        /// Get all providers from the database.
+        /// </summary>
+        /// <returns>List of all providers.</returns>
+        [HttpGet]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        public async Task<ActionResult<IEnumerable<Provider>>> GetProviders()
+        {
+            return Ok(await providerService.GetAll().ConfigureAwait(false));
+        }
+
+        /// <summary>
+        /// Get provider by it's key.
+        /// </summary>
+        /// <param name="id">The key in the database.</param>
+        /// <returns>Provider element with some id.</returns>
+        [HttpGet("{id}")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        public async Task<ActionResult<ProviderDto>> GetProviderById(long id)
+        {
+            if (id == 0)
+            {
+                return BadRequest("Id cannot be 0.");
+            }
+
+            return Ok(await providerService.GetById(id).ConfigureAwait(false));
+        }
+
+        /// <summary>
+        /// Create new provider.
+        /// </summary>
+        /// <param name="providerDTO">Element which must be added.</param>
+        /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
+        [Authorize(Roles = "provider,admin")]
+        [HttpPost]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status201Created)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        public async Task<ActionResult<Provider>> Create(ProviderDto providerDTO)
+        {
+            if (!ModelState.IsValid)
+            {
+                return BadRequest(ModelState);
+            }
+
+            if (providerDTO == null)
+            {
+                return BadRequest(providerDTO);
```

Comment:
does any error message needed here?

Reply:
StyleCop asks for null check

Reply:
will user understand what went wrong in this case?